src/zcash/Note.cpp

```diff
@@ -281,13 +362,44 @@ boost::optional<SaplingNotePlaintext> SaplingNotePlaintext::decrypt(
         return boost::none;
     }
 
-    return ret;
+    if (plaintext.get_leadbyte() == 0x02) {
```

Comment:
I believe we want to perform this check all the time, since we always have `esk` (before and after Canopy activation) in the outgoing viewing key context and this check is mandated by the existing protocol specification. So, let's remove this if statement and just always perform this check.

Reply:
for `v1` plaintexts, i think the check `(esk == plaintext.generate_esk())` would fail because `generate_esk()` would pick another random `esk`.

Reply:
Oh yeah, you're right. Okay.

Reply:
Yes, the current spec is actually wrong on this; I had just changed it in my local copy. @therealyingtong is correct. The new spec will use `!= 0x01` though (it makes no difference for the time being, but might if we add a new version in future).

```suggestion
    if (plaintext.get_leadbyte() != 0x01) {
```

Reply:
changed to `if (plaintext.get_leadbyte() != 0x01) {`